tests/unit/s2n_shutdown_test.c

```diff
@@ -0,0 +1,122 @@
+            EXPECT_NOT_NULL(config = s2n_config_new());
+            struct s2n_connection *conn;
+            EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_CLIENT));
+            EXPECT_SUCCESS(s2n_connection_set_config(conn, config));
+
+            /* Verify state prior to alert */
+            EXPECT_FALSE(conn->did_recv_close_notify);
+
+            /* Write and process the alert */
+            EXPECT_SUCCESS(s2n_stuffer_write_bytes(&conn->in, close_notify_alert, sizeof(close_notify_alert)));
+            EXPECT_SUCCESS(s2n_process_alert_fragment(conn));
+
+            /* Verify state after alert */
+            EXPECT_TRUE(conn->did_recv_close_notify);
+
+            s2n_blocked_status blocked;
+            EXPECT_SUCCESS(s2n_shutdown(conn, &blocked));
+
+            EXPECT_SUCCESS(s2n_connection_free(conn));
+            EXPECT_SUCCESS(s2n_config_free(config));
+        }
+
+        /* Await close_notify if did_recv_close_notify is not set */
+        {
+            struct s2n_config *config;
+            EXPECT_NOT_NULL(config = s2n_config_new());
+            struct s2n_connection *conn;
+            EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_CLIENT));
+            EXPECT_SUCCESS(s2n_connection_set_config(conn, config));
+
+            /* Set mock send impl */
+            EXPECT_SUCCESS(s2n_connection_set_send_cb(conn, &mock_send_impl));
+
+            /* Verify state prior to alert */
+            EXPECT_FALSE(conn->did_recv_close_notify);
+
+            s2n_blocked_status blocked;
+            EXPECT_FAILURE_WITH_ERRNO(s2n_shutdown(conn, &blocked), S2N_ERR_IO);
+            EXPECT_EQUAL(blocked, S2N_BLOCKED_ON_READ);
+
+            /* Verify state after shutdown attempt */
+            EXPECT_FALSE(conn->did_recv_close_notify);
+
+            EXPECT_SUCCESS(s2n_connection_free(conn));
+            EXPECT_SUCCESS(s2n_config_free(config));
+        }
+
+        /* Do not await close_notify if did_recv_close_notify is set */
+        {
+            struct s2n_config *config;
+            EXPECT_NOT_NULL(config = s2n_config_new());
+            struct s2n_connection *conn;
+            EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_CLIENT));
+            EXPECT_SUCCESS(s2n_connection_set_config(conn, config));
+
+            /* Set mock send impl */
+            EXPECT_SUCCESS(s2n_connection_set_send_cb(conn, &mock_send_impl));
+
+            /* Verify state prior to alert */
+            EXPECT_FALSE(conn->did_recv_close_notify);
+
+            /* Write and process the alert */
+            EXPECT_SUCCESS(s2n_stuffer_write_bytes(&conn->in, close_notify_alert, sizeof(close_notify_alert)));
+            EXPECT_SUCCESS(s2n_process_alert_fragment(conn));
+
+            /* Verify state after alert */
+            EXPECT_TRUE(conn->did_recv_close_notify);
+
+            s2n_blocked_status blocked;
+            EXPECT_SUCCESS(s2n_shutdown(conn, &blocked));
+            EXPECT_EQUAL(blocked, S2N_NOT_BLOCKED);
```

Comment:
If s2n_shutdown succeeds, is blocked guaranteed to be S2N_NOT_BLOCKED?

Reply:
yes. blocked gets used/set from withing s2n_shutdown by two functions `s2n_flush` and `s2n_recv_close_notify`. both set `blocked` to S2N_NOT_BLOCKED on success (which causes s2n_shutdown to be successful).